### ISSUE TITLE
Suggestion: Better handling of ffmpeg exit codes

### ISSUE THREAD

**Issue by tilllt** [jellyfin/jellyfin#2314]
After some problems with my new Radeon rx550, i got vaapi running, so i passed the device into my jellyfin docker and according to the logs, jellyfin uses the HW transcoding by the card.

First i ignored the critics that warned against using HW transcoding, since jellyfin apparently doesnt fall-back to SW transcoding if HW transcoding fails:
https://www.reddit.com/r/jellyfin/comments/e7ydga/amd_radeon_for_transcoding_buying_advice/fb3ii8t/

But now i have multiple media files that dont work work with HW transcoding, so i understand why this is a problem...

![Screenshot_14](https://user-images.githubusercontent.com/1854364/72722288-3e939980-3b7e-11ea-8337-496b1b00112c.jpg)

So i wonder, whats the state of proper ffmpeg exit code handling to silently switch between HW & SW transcoding? is there something in the works?

this is a mediainfo output of the video stream of a clip that does not play using HW encoding using vaapi: 

```
Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High@L4
Format settings                          : CABAC / 4 Ref Frames
Format settings, CABAC                   : Yes
Format settings, RefFrames               : 4 frames
Codec ID                                 : V_MPEG4/ISO/AVC
Duration                                 : 1 h 5 min
Bit rate                                 : 10.2 Mb/s
Width                                    : 1 920 pixels
Height                                   : 960 pixels
Display aspect ratio                     : 2.000
Frame rate mode                          : Constant
Frame rate                               : 23.976 (24000/1001) FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Bits/(Pixel*Frame)                       : 0.230
Stream size                              : 4.62 GiB (96%)
Language                                 : English
Default                                  : Yes
Forced                                   : No
Color range                              : Limited
Color primaries                          : BT.709
Transfer characteristics                 : BT.709
Matrix coefficients                      : BT.709
```

**Comment by tilllt** [comment]
And a semi-related question: can / does jellyfin transcode for a chromecast, to make sure only compatible streams are played on the chromecast?

**Comment by JustAMan** [comment]
> can / does jellyfin transcode for a chromecast, to make sure only compatible streams are played on the chromecast?

Yes, though there's a separate issue submitted that transcode could potentially fail.

On your main issue - this has nothing to do with ffmpeg exit code handling, as there's no logic to fallback currently. So this is more like a feature request or improvement...

**Comment by tilllt** [comment]
I dont quite understand the reply, 

"this has nothing to do with ffmpeg exit code handling, as there's no logic to fallback currently."

From the [FFMPEG Docs:](https://trac.ffmpeg.org/wiki/HWAccelIntro)

>  Internal hwaccel decoders are enabled via the -hwaccel option. The software decoder starts normally, but if it detects a stream which is decodable in hardware then it will attempt to delegate all significant processing to that hardware. If the stream is not decodable in hardware (for example, it is an unsupported codec or profile) then it will still be decoded in software automatically. If the hardware requires a particular device to function (or needs to distinguish between multiple devices, say if several graphics cards are available) then one can be selected using -hwaccel_device.
> 
> External wrapper decoders are used by setting a specific decoder with the -codec:v option. Typically they are named codec_api (for example: h264_cuvid). These decoders require the codec to be known in advance, and do not support any fallback to software if the stream is not supported.

**Comment by JustAMan** [comment]
You're talking about fallback logic in ffmpeg, which is irrelevant because all the processing is done in Jellyfin itself (including setting the codecs). Thus it is not relevant to handling exit codes per se, but relevant to handling decode issues and falling back to software decoding. Maybe it would be enough to just use `-hwaccel` properly, maybe not - ffmpeg CLI isn't the easiest thing, and right now the way Jellyfin handles it is suboptimal at best (or, rather, pretty poor IMHO). But again, this has nothing to do with exit codes themselves :)

What you seem to _actually_ want is this - https://features.jellyfin.org/posts/318/gracefully-fallback-on-software-encoding-when-hardware-encoding-crashes

If I'm right I suggest you closing this issue and upvoting that feature request instead.

**Comment by tilllt** [comment]
I will happily upvote the Feature Suggestion, i just dont agree its a Feature. In my understanding it is a bug, either HW transcoding is supported or not. If Jellyfin fails to correctly choose a working decoder although the host system offers one, thats not really "optional", it completely breaks the possibility to use HW transcoding unless you have all your media, which i guess nobody has. Dont get me wrong, i think Jellyfin is a wonderful project and i will happily disable HW Transcoding for the time being. But saying its a "feature" to pass a video stream to a correct decoder is a bit weird imho.
